monkey/monkey_island/cc/repository/i_machine_repository.py

```diff
@@ -1,20 +1,74 @@
-from abc import ABC
-from typing import Optional, Sequence
+from abc import ABC, abstractmethod
+from ipaddress import IPv4Address
+from typing import Sequence
+
+from common.types import HardwareID
+from monkey_island.cc.models import Machine, MachineID
 
 
 class IMachineRepository(ABC):
-    # TODO define Machine object(ORM model)
-    def save_machine(self, machine: Machine):  # noqa: F821
-        pass
-
-    # TODO define Machine object(ORM model)
-    # TODO define or re-use machine state.
-    # TODO investigate where should the state be stored in edge or both edge and machine
-    def get_machines(
-        self,
-        id: Optional[str] = None,
-        ips: Optional[Sequence[str]] = None,
-        state: Optional[MachineState] = None,  # noqa: F821
-        is_island: Optional[bool] = None,  # noqa: F841
-    ) -> Sequence[Machine]:  # noqa: F821
-        pass
+    """A repository used to store and retrieve Machines"""
+
+    @abstractmethod
+    def create_machine(self) -> Machine:
+        """
+        Create a new `Machine` in the repository
+
+        :return: A new `Machine` with a unique ID
+        :raises StorageError: If a new `Machine` could not be created
+        """
+
+    @abstractmethod
+    def update_machine(self, machine: Machine):
+        """
+        Update an existing `Machine` in the repository
+
+        :param machine: An updated Machine object to store in the repository
+        :raises UnknownRecordError: If the provided `Machine` does not exist in the repository
+        :raises StorageError: If an error occurred while attempting to store the `Machine`
+        """
+
+    @abstractmethod
+    def get_machine_by_id(self, machine_id: MachineID) -> Machine:
+        """
+        Get a `Machine` by ID
+
+        :param machine_id: The ID of the `Machine` to be retrieved
+        :return: A `Machine` with a matching `id`
+        :raises UnknownRecordError: If a `Machine` with the specified `id` does not exist in the
+                                    repository
+        :raises RetrievalError: If an error occurred while attempting to retrieve the `Machine`
+        """
+
+    @abstractmethod
+    def get_machine_by_hardware_id(self, hardware_id: HardwareID) -> Machine:
+        """
+        Get a `Machine` by ID
+
+        :param hardware_id: The hardware ID of the `Machine` to be retrieved
+        :return: A `Machine` with a matching `hardware_id`
+        :raises UnknownRecordError: If a `Machine` with the specified `hardware_id` does not exist
+                                    in the repository
+        :raises RetrievalError: If an error occurred while attempting to retrieve the `Machine`
+        """
+
+    @abstractmethod
+    def get_machines_by_ip(self, ip: IPv4Address) -> Sequence[Machine]:
```

Comment:
Consider taking a list of IP's, since it gives more flexibility to the caller

Reply:
It also adds some ambiguity. Should that list of IPs be `and` or `or`? We can choose and then document it, but what if we want the other one?

If we need that capability, we can add it when we need it. Until then, I claim YAGNI.